src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py

```diff
@@ -0,0 +1,471 @@
+                    with torch.no_grad():
+                        for k in range(len(candidate_ops)):
+                            if k != active_id:
+                                out_k = candidate_ops[k](_x.data)
+                            else:
+                                out_k = _output.data
+                            grad_k = torch.sum(out_k * grad_output)
+                            binary_grads[k] = grad_k
+                    return binary_grads
+                return backward
+            output = ArchGradientFunction.apply(
+                x, self.AP_path_wb, run_function(mutable.key, mutable.choices, self.active_index[0]),
+                backward_function(mutable.key, mutable.choices, self.active_index[0], self.AP_path_wb))
+        else:
+            output = self.active_op(mutable)(x)
+        return output
+
+    @property
+    def probs_over_ops(self):
+        """
+        Apply softmax on alpha to generate probability distribution
+
+        Returns
+        -------
+        pytorch tensor
+            probability distribution
+        """
+        probs = F.softmax(self.AP_path_alpha, dim=0)  # softmax to probability
+        return probs
+
+    @property
+    def chosen_index(self):
+        """
+        choose the op with max prob
+
+        Returns
+        -------
+        int
+            index of the chosen one
+        numpy.float32
+            prob of the chosen one
+        """
+        probs = self.probs_over_ops.data.cpu().numpy()
+        index = int(np.argmax(probs))
+        return index, probs[index]
+
+    def active_op(self, mutable):
+        """
+        assume only one path is active
+
+        Returns
+        -------
+        PyTorch module
+            the chosen operation
+        """
+        return mutable.choices[self.active_index[0]]
+
+    @property
+    def active_op_index(self):
+        """
+        return active op's index, the active op is sampled
+
+        Returns
+        -------
+        int
+            index of the active op
+        """
+        return self.active_index[0]
+
+    def set_chosen_op_active(self):
+        """
+        set chosen index, active and inactive indexes
+        """
+        chosen_idx, _ = self.chosen_index
+        self.active_index = [chosen_idx]
+        self.inactive_index = [_i for _i in range(0, chosen_idx)] + \
+                              [_i for _i in range(chosen_idx + 1, self.n_choices)]
+
+    def binarize(self, mutable):
+        """
+        Sample based on alpha, and set binary weights accordingly.
+        AP_path_wb is set in this function, which is called binarize.
+
+        Parameters
+        ----------
+        mutable : LayerChoice
+            this layer's mutable
+        """
+        self.log_prob = None
+        # reset binary gates
+        self.AP_path_wb.data.zero_()
+        probs = self.probs_over_ops
+        if MixedOp.forward_mode == 'two':
```

Comment:
???? poor code readability

Reply:
which part do you mean? 'two'? or forward_mode?

Reply:
what're the "two" forward_mode means?  what's kind of forward_mode are "two" forward_mode?

Reply:
added the following in `forward` docstring:

The 'two' mode is explained in section 3.2.1 in the paper.
The 'full_v2' mode is explained in Appendix D in the paper.